File: solutions/python/anagram/anagram.py
"""Find the subset of the candidates that are anagrams."""


def find_anagrams(subject: str, candidates: list[str]) -> list[str]:
    """Returns all candidates that are anagrams of,
    but not equal to, 'subject' word.

    :param subject: str - the word to check
    :param candidates: list[str] - potential matches
    :return: list[str] - anagrams of 'subject'
    """
    base = subject.lower()
    sorted_base = sorted(base)

    def is_anagram(candidate: str) -> bool:
        if len(candidate) != len(base):
            return False
        word = candidate.lower()
        if word == base:
            return False
        return sorted(word) == sorted_base

    return [word for word in candidates if is_anagram(word)]

File: solutions/javascript/anagram/anagram.js
// @ts-check

/**
 * Returns all candidates that are anagrams of, but not equal to, 'target'.
 *
 * @param{string} target
 * @param{string[]} candidates
 *
 * @returns{string[]} anagrams
 */
export const findAnagrams = (target, candidates) => {
  const base = target.toLowerCase();
  const anagrams = [];

  const sorted = (/** @type {string} */ s) => s.split("").sort().join("");

  for (const candidate of candidates) {
    const word = candidate.toLowerCase();
    if (base === word) continue;
    if (sorted(base) !== sorted(word)) continue;
    anagrams.push(candidate);
  }
  return anagrams;
};
